### ISSUE TITLE
[ADE LVL 2] Page change / loading not announced

### ISSUE THREAD

**Issue by outerpress** [department-of-veterans-affairs/va.gov-team#111661]
> [!IMPORTANT]
> Review duplicate issue before starting work: https://github.com/department-of-veterans-affairs/va.gov-team/issues/111489

## Description
 
On some pages, a loading message appears that isn't announced to screenreaders. For example, on the facility page the message appears: \- “We’re checking if we can create an appointment for you at this facility…”  

This message doesn’t announce to screen reader users, who don’t hear any announcement until the page is finished loading. No “page change in progress,” “loading,” or similar messages.  

[Clip](https://coformaco.zoom.us/clips/share/-9g-r_63StyHCB0S8ZVvRQ) showing two loading screens:  
    1. Going between type of care and facility, where it announces “Page change in progress”  
    2. Going between facility and clinic, where nothing is announced  

* **Pages affected:** Going between:  
  * [What type of care do you need?](https://staging.va.gov/my-health/appointments/schedule/type-of-care) and  [Which VA facility would you like to go to?](https://staging.va.gov/my-health/appointments/schedule/location)  
  *  [Which VA facility would you like to go to?](https://staging.va.gov/my-health/appointments/schedule/location) and [Which VA clinic would you like to go to?](https://staging.va.gov/my-health/appointments/schedule/clinic)   
  * Would you like to cancel this appointment? and You have canceled your appointment  
  * [Which VA facility would you like to go to?](https://staging.va.gov/my-health/appointments/schedule/location)  and [When would you like an appointment (VA request)?](https://staging.va.gov/my-health/appointments/schedule/va-request/)  
  * [Review and submit your request (VA request)](https://staging.va.gov/my-health/appointments/schedule/va-request/review) and We have received your request  
  * Would you like to cancel this request? and You have canceled your request  
* **Device:** JAWS, Mac VO  
* **Orientation:** Both  
* **Expected severity level:** 2  
* **Expected experience standard:** User can perceive all meaningful elements  
* **How to recreate:**  
  * Turn on a screen reader (I used JAWS)  
  * Pick a facility on [Which VA facility would you like to go to?](https://staging.va.gov/my-health/appointments/schedule/location) and select “Continue”  
* **Recommendation:** If there’s a loading screen like this, a message must be conveyed to screen reader users. Since there’s actual text on screen and not just a loading indicator, that text should be communicated to assistive tech. Use the [VA loading indicator](https://design.va.gov/storybook/?path=/docs/components-va-loading-indicator--docs):  
  * Set “label” to duplicate the “message” prop  
  * Use the “set-focus” prop  
  * This will move focus to the loading indicator div on load and announce the aria-label (set in “label”).

---

## Development Checklist
Background: the `toggle` toggle is on (remove if not needed) 

- [ ] Task 1
- [ ] Task 2

### Design Assets
- [Figma file]() (put N/A if not needed) 

### Developer Reference

- [Swagger File]() (remove if not needed)
- [File path]() (remove if not needed)

## Definition of Done
- [ ] All tasks criteria are met
- [ ] [Technical Documentation]() is updated (must confirm if needed/or remove) 
- [ ] [Feature Reference (GitHub)](https://github.com/department-of-veterans-affairs/va.gov-team/tree/master/products/health-care/appointments/va-online-scheduling/feature-reference) is updated
- [ ] UX - [Feature Reference (Figma)](https://www.figma.com/design/eonNJsp57eqfPqx7ydsJY9/Feature-Reference-%7C-Appointments-FE?node-id=1150-105402&t=zRNPBqgPDVU0dDP8-1) is updated

**Comment by StevenStraily** [comment]
@coforma-jamie There was a question about this, and we could use your expertise!
I remember doing something similar to this on DBEX, and we weren't changing the focus and reading off the text verbatim. I think we used describedby or polite, can't remember, for a custom loading message that fit, so we wouldn't end up spamming them every time the message changed quickly to represent a different endpoint being hit.

Would you advise that we do something similar here? The alternative is that they will likely get bombarded by a few different messages within a couple of seconds on some of these loading screens.

**Comment by coforma-jamie** [comment]
Ooh! That's a great point. I'm a little torn actually, because there are two schools of thought:
- provide an equivalent experience
- provide the most usable experience

For example, in this case, the most usable _and_ equivalent experience would be an announcement like "Page loading. Finding available locations for your appointment," because it conveys the action represented by the graphic and the exact text that sighted users see:
<img width="344" height="123" alt="Image" src="https://github.com/user-attachments/assets/3d643c36-6e22-4dfc-a843-ef315b521887" />

But if someone had to go through three loading screens before the appointments actually loaded, and each screen had its own message? Then there's a conflict.
- It's more _usable_ (less frustrating) to just hear one message, since the individual messages are duplicative and might get cut off depending on recitation speed...
- but it's more _equivalent_ to hear each announcement because that's what's happening to sighted users too. Limiting announcements for one type of user, but not another, isn't an equivalent experience.

I'm going to pose this question to other VA accessibility folks and see if there's a consensus. But wanted you to understand my reasoning first :)

**Comment by coforma-jamie** [comment]
Spoke to another specialist, who wisely said "No need to replicate a bad UX pattern for the sake of providing an equivalent experience. I would also work to address the UX pattern."

So in this case, as long as you're conveying the intent ("loading available locations" or whatnot), it should be fine to announce it like you did with DBEX <3

**Comment by StevenStraily** [comment]
Awesome, thanks, Jamie!

**Comment by eselkin** [comment]
So, in the recommendations, there's the issue that when `set-focus` is used on the `va-loading-indicator` SRs make an outline around the indicator. This outline and focus was requested to be removed in another PR about a different indicator. However, maybe we shouldn't be removing the focus or the outline, but changing the outline or changing something about the outline when set-focus is used?

**Comment by eselkin** [comment]
It seems the page loading readout to SR takes place on the button action for navigation to the next form page. However, some pages read data before the load so the navigating to next page information gets heard. However, for form pages that don't load data the navigation is "instant" so it cuts off the VO (only one I can test) and so you don't hear that preparing to navigate information.

**Comment by eselkin** [comment]
Ok, so clarifying the Page change in progress announcements

1. When the page change is too fast, no SR announces the Page change in progress notification (this is the same on VO, JAWS, and NVDA)
2. When the button is disabled in VoiceOver only the system makes an announcement about the button being dimmed. The only way to get it not to say that is to add `aria-disable="true"` while disabled. However, then VoiceOver announces something else. Unfortunately, no other element even with `aria-live="assertive"` and `role="alert"` makes any announcement to the page. It DOES make the announcement on NVDA and JAWS.
3. If you load a component in react dynamically, SR don't really announce that if the content changed is not just plain text. So, you can just change the text via JavaScript and this does make the announcement. However, this does not overcome number 2 above.

The only page due to processing time that may announce the page change in progress (only JAWS and NVDA) is really the first page (choosing type-of-care).

@outerpress and I previously discussed that we can slow down the page changes but this seems like a bad user experience for all and cannot be selective to just screen readers.

I tried many many alternatives for how and when to update the text content to try to trigger the Page change notification to SRs but none overcame the VO issue for the first page.

**Comment by coforma-jamie** [comment]
Reading through the comments - there are a few things I want to point out about default screen reader behavior:

> When the page change is too fast, no SR announces the Page change in progress notification (this is the same on VO, JAWS, and NVDA)
If the page change is so fast that it doesn't announce the progress notification, that should be ok? The notification is meant to tell folks that something is loading, if it's loaded already then there's nothing to announce. 

> when set-focus is used on the va-loading-indicator SRs make an outline around the indicator
That's expected behavior - by default, an SR will show an indicator around whatever the virtual cursor is focused on. Some SRs actually let you turn that off! 

> VoiceOver only the system makes an announcement about the button being dimmed
"Dimmed" is how VoiceOver refers to disabled fields. "Dimmed" is definitely a confusing word, but it's also the standard (try it on any disabled button) and would be expected by experienced VoiceOver users.

**Comment by eselkin** [comment]
@coforma-jamie I wasn't really worried about the dimmed (I read up on Dimmed). It's just that VoiceOver announced it where the others didn't. The issue was that nothing can interrupt that or announce something after that has occurred (because the description takes a long time), so any plan you make to announce the page change on the other screen readers can't be made on VoiceOver. There are occasional spots where VO does announce it, but it seems random.

**Comment by outerpress** [comment]
Sorry for the delay @eselkin - this sounds great to me! Validated 👍

**Comment by JunTaoLuo** [comment]
Oops, I missed the validated comment. There's no continuous deployment at the moment so the daily deploys are all we need. This has been deployed for a while so I'm closing this. Thanks all!
